File: packages/idea-core/src/idea_core/contracts/catalog.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator, FormatChecker
from referencing import Registry, Resource

from idea_core.contracts.validate import DEFAULT_CONTRACT_DIR, OPENRPC_FILE


class ContractRuntimeError(RuntimeError):
    """Runtime contract validation error."""


@dataclass(frozen=True)
class MethodContract:
    name: str
    params: list[dict[str, Any]]
    result_schema: dict[str, Any]


class ContractCatalog:
    def __init__(self, contract_dir: Path = DEFAULT_CONTRACT_DIR) -> None:
        self.contract_dir = contract_dir
        self.openrpc_path = (contract_dir / OPENRPC_FILE).resolve()
        self.openrpc = self._load_json(self.openrpc_path)
        self.contract_version = str(self.openrpc.get("info", {}).get("version", "unknown"))
        self.registry = self._build_registry(contract_dir)
        self.format_checker = FormatChecker()
        self.methods: dict[str, MethodContract] = {}

        for method in self.openrpc.get("methods", []):
            self.methods[method["name"]] = MethodContract(
                name=method["name"],
                params=method.get("params", []),
                result_schema=method.get("result", {}).get("schema", {}),
            )

        error_contract = self.openrpc.get("x-error-data-contract", {})
        self.error_data_schema = error_contract.get("schema")

    @staticmethod
    def _load_json(path: Path) -> dict[str, Any]:
        return json.loads(path.read_text(encoding="utf-8"))

    def _build_registry(self, contract_dir: Path) -> Registry:
        registry = Registry()
        for schema_path in sorted(contract_dir.glob("*.schema.json")):
            schema = self._load_json(schema_path)
            registry = registry.with_resource(
                schema_path.resolve().as_uri(), Resource.from_contents(schema)
            )
        return registry

    def _validate_with_schema(
        self,
        schema: dict[str, Any],
        instance: Any,
        base_uri: str,
    ) -> None:
        wrapped_schema = {
            "$schema": "https://json-schema.org/draft/2020-12/schema",
            "$id": base_uri,
            **schema,
        }
        validator = Draft202012Validator(
            wrapped_schema,
            registry=self.registry,
            format_checker=self.format_checker,
        )
        errors = sorted(validator.iter_errors(instance), key=lambda e: str(e.path))
        if errors:
            first = errors[0]
            location = "/".join(str(part) for part in first.path)
            raise ContractRuntimeError(
                f"schema_invalid at '{location or '<root>'}': {first.message}"
            )

    def validate_request_params(self, method_name: str, params: dict[str, Any]) -> None:
        if method_name not in self.methods:
            raise ContractRuntimeError(f"unknown method contract: {method_name}")
        if not isinstance(params, dict):
            raise ContractRuntimeError("params must be an object (by-name)")

        method = self.methods[method_name]
        required_params = {p["name"] for p in method.params if p.get("required")}

        missing = sorted(required_params - params.keys())
        if missing:
            raise ContractRuntimeError(f"missing required params: {', '.join(missing)}")

        allowed = {p["name"] for p in method.params}
        extra = sorted(set(params.keys()) - allowed)
        if extra:
            raise ContractRuntimeError(f"unknown params: {', '.join(extra)}")

        for param in method.params:
            name = param["name"]
            if name not in params:
                continue
            schema = param.get("schema")
            if not isinstance(schema, dict):
                continue
            base_uri = self.openrpc_path.resolve().as_uri() + f"#/{method_name}/params/{name}"
            self._validate_with_schema(schema, params[name], base_uri=base_uri)

    def validate_result(self, method_name: str, result: dict[str, Any]) -> None:
        method = self.methods[method_name]
        schema = method.result_schema
        base_uri = self.openrpc_path.resolve().as_uri() + f"#/{method_name}/result"
        self._validate_with_schema(schema, result, base_uri=base_uri)

    def validate_against_ref(self, ref: str, instance: Any, *, base_name: str) -> None:
        schema = {"$ref": ref}
        base_uri = self.openrpc_path.resolve().as_uri() + f"#/{base_name}"
        self._validate_with_schema(schema, instance, base_uri=base_uri)

    def validate_error_data(self, error_data: dict[str, Any]) -> None:
        if not isinstance(self.error_data_schema, dict):
            return
        base_uri = self.openrpc_path.resolve().as_uri() + "#/x-error-data-contract/schema"
        self._validate_with_schema(self.error_data_schema, error_data, base_uri=base_uri)

File: packages/idea-engine/src/contracts/catalog.ts
import { readdirSync, readFileSync } from 'fs';
import { resolve } from 'path';
import { pathToFileURL } from 'url';
import Ajv2020 from 'ajv/dist/2020.js';
import { DEFAULT_CONTRACT_DIR, type OpenRpcMethod, type OpenRpcDocument } from './openrpc.js';

type AjvValidator = ((value: unknown) => boolean) & {
  errors?: Array<{ instancePath?: string; message?: string }>;
};

type AjvInstance = {
  addFormat: (name: string, validate: (value: string) => boolean) => void;
  addSchema: (schema: Record<string, unknown>, key?: string) => void;
  compile: (schema: Record<string, unknown>) => AjvValidator;
};

type AjvConstructor = new (options: Record<string, unknown>) => AjvInstance;

const Ajv2020Ctor = Ajv2020 as unknown as AjvConstructor;

export class ContractRuntimeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ContractRuntimeError';
  }
}

function formatValidationError(errors: AjvValidator['errors']): string {
  const first = errors?.[0];
  if (!first) {
    return "schema_invalid at '<root>': validation failed";
  }
  const location = first.instancePath ? first.instancePath.slice(1) || '<root>' : '<root>';
  return `schema_invalid at '${location}': ${first.message ?? 'validation failed'}`;
}

function isUuid(value: string): boolean {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value);
}

function isDateTime(value: string): boolean {
  return typeof value === 'string' && value.includes('T') && !Number.isNaN(Date.parse(value));
}

function isUri(value: string): boolean {
  try {
    const parsed = new URL(value);
    return parsed.protocol.length > 0;
  } catch {
    return false;
  }
}

export class IdeaEngineContractCatalog {
  readonly contractDir: string;
  readonly contractVersion: string;
  private readonly ajv: AjvInstance;
  private readonly methods = new Map<string, OpenRpcMethod>();
  private readonly openrpcUri: string;
  private readonly errorDataSchema?: Record<string, unknown>;

  constructor(contractDir = DEFAULT_CONTRACT_DIR) {
    this.contractDir = resolve(contractDir);
    const openrpcPath = resolve(this.contractDir, 'idea_core_rpc_v1.openrpc.json');
    this.openrpcUri = pathToFileURL(openrpcPath).href;
    const openrpc = JSON.parse(readFileSync(openrpcPath, 'utf8')) as OpenRpcDocument;
    this.contractVersion = String(openrpc.info?.version ?? 'unknown');
    this.ajv = new Ajv2020Ctor({
      allErrors: true,
      strict: false,
      validateFormats: true,
      addUsedSchema: false,
    });
    this.ajv.addFormat('uuid', isUuid);
    this.ajv.addFormat('date-time', isDateTime);
    this.ajv.addFormat('uri', isUri);
    this.loadSchemas();

    for (const method of openrpc.methods ?? []) {
      this.methods.set(method.name, method);
    }
    this.errorDataSchema = openrpc['x-error-data-contract']?.schema;
  }

  validateRequestParams(methodName: string, params: unknown): void {
    const method = this.methods.get(methodName);
    if (!method) {
      throw new ContractRuntimeError(`unknown method contract: ${methodName}`);
    }
    if (!params || typeof params !== 'object' || Array.isArray(params)) {
      throw new ContractRuntimeError('params must be an object (by-name)');
    }

    const record = params as Record<string, unknown>;
    const required = new Set((method.params ?? []).filter(param => param.required).map(param => param.name));
    const missing = [...required].filter(name => !(name in record));
    if (missing.length > 0) {
      throw new ContractRuntimeError(`missing required params: ${missing.join(', ')}`);
    }

    const allowed = new Set((method.params ?? []).map(param => param.name));
    const extras = Object.keys(record).filter(name => !allowed.has(name));
    if (extras.length > 0) {
      throw new ContractRuntimeError(`unknown params: ${extras.join(', ')}`);
    }

    for (const param of method.params ?? []) {
      if (!(param.name in record) || !param.schema || typeof param.schema !== 'object') {
        continue;
      }
      this.validateWithSchema(
        param.schema as Record<string, unknown>,
        record[param.name],
        this.scopedUri(`${methodName}/params/${param.name}`),
      );
    }
  }

  validateResult(methodName: string, result: unknown): void {
    const method = this.methods.get(methodName);
    if (!method?.result?.schema || typeof method.result.schema !== 'object') {
      throw new ContractRuntimeError(`unknown result contract: ${methodName}`);
    }
    this.validateWithSchema(method.result.schema, result, this.scopedUri(`${methodName}/result`));
  }

  validateAgainstRef(ref: string, instance: unknown, baseName: string): void {
    this.validateWithSchema({ $ref: ref }, instance, this.scopedUri(baseName));
  }

  validateErrorData(errorData: Record<string, unknown>): void {
    if (!this.errorDataSchema) {
      return;
    }
    this.validateWithSchema(this.errorDataSchema, errorData, this.scopedUri('x-error-data-contract/schema'));
  }

  private loadSchemas(): void {
    for (const entry of readdirSync(this.contractDir).filter(name => name.endsWith('.schema.json')).sort()) {
      const path = resolve(this.contractDir, entry);
      const schema = JSON.parse(readFileSync(path, 'utf8')) as Record<string, unknown>;
      this.ajv.addSchema({ ...schema, $id: pathToFileURL(path).href }, pathToFileURL(path).href);
    }
  }

  private validateWithSchema(schema: Record<string, unknown>, instance: unknown, baseUri: string): void {
    const validator = this.ajv.compile({
      $schema: 'https://json-schema.org/draft/2020-12/schema',
      $id: baseUri,
      ...schema,
    });
    if (!validator(instance)) {
      throw new ContractRuntimeError(formatValidationError(validator.errors));
    }
  }

  private scopedUri(scope: string): string {
    return `${this.openrpcUri}?scope=${encodeURIComponent(scope)}`;
  }
}
